backend/ethereum/channel/subscription.go

```diff
@@ -17,115 +17,74 @@ package channel
+		err:    make(chan error, 1),
 	}
-
-	// Start event updater routine
 	go rsub.updateNext(ctx, events, a)
 
-	// find past event, if any
-	var ev *adjudicator.AdjudicatorChannelUpdate
-	for iter.Next() {
-		ev = iter.Event // fast-forward to newest event
-	}
-	// nolint:errcheck,gosec,gosec
-	iter.Close()
-	if err := iter.Error(); err != nil {
-		err = cherrors.CheckIsChainNotReachableError(err)
-		sub.Unsubscribe()
-		return nil, errors.WithMessage(err, "event iterator")
-	}
-	// Pass non-nil past event to updater
-	if ev != nil {
-		rsub.past = true
-		events <- ev
-	}
-
 	return rsub, nil
 }
 
-// filterWatch sets up a filter and a subscription on events.
-func (a *Adjudicator) filterWatch(ctx context.Context, events chan *adjudicator.AdjudicatorChannelUpdate, params *channel.Params) (sub event.Subscription, iter *adjudicator.AdjudicatorChannelUpdateIterator, err error) {
-	defer func() {
-		if err != nil && sub != nil {
-			sub.Unsubscribe()
-		}
-	}()
-	// Watch new events
-	watchOpts, err := a.NewWatchOpts(ctx)
-	if err != nil {
-		return nil, nil, errors.WithMessage(err, "creating watchopts")
-	}
-	sub, err = a.contract.WatchChannelUpdate(watchOpts, events, []channel.ID{params.ID()})
-	if err != nil {
-		err = cherrors.CheckIsChainNotReachableError(err)
-		return nil, nil, errors.WithMessagef(err, "watching events")
-	}
-
-	// Filter old Events
-	filterOpts, err := a.NewFilterOpts(ctx)
-	if err != nil {
-		return nil, nil, errors.WithMessage(err, "creating filter opts")
-	}
-	iter, err = a.contract.FilterChannelUpdate(filterOpts, []channel.ID{params.ID()})
-	if err != nil {
-		err = cherrors.CheckIsChainNotReachableError(err)
-		return nil, nil, errors.WithMessage(err, "filtering events")
-	}
-
-	return sub, iter, nil
-}
-
-// RegisteredSub implements the channel.RegisteredSubscription interface.
+// RegisteredSub implements the channel.AdjudicatorSubscription interface.
 type RegisteredSub struct {
-	cr   ethereum.ChainReader          // chain reader to read block time
-	sub  event.Subscription            // Event subscription
-	next chan channel.AdjudicatorEvent // Event sink
-	err  chan error                    // error from subscription
-	past bool                          // whether there was a past event when the subscription was created
+	cr     ethereum.ChainReader   // chain reader to read block time
+	sub    *subscription.EventSub // Event subscription
+	subErr chan error
+	next   chan channel.AdjudicatorEvent // Event sink
+	err    chan error                    // error from subscription
+	closed sync.Once
 }
 
-// HasPast indicates whether there was a past event when the subscription was created.
-func (r *RegisteredSub) HasPast() bool {
-	return r.past
-}
-
-func (r *RegisteredSub) updateNext(ctx context.Context, events chan *adjudicator.AdjudicatorChannelUpdate, a *Adjudicator) {
+// nolint: gocognit
+func (r *RegisteredSub) updateNext(ctx context.Context, events chan *subscription.Event, a *Adjudicator) {
```

Comment:
Why do we need such a complicated logic again? Couldn't we just use our subscription?

Reply:
I dont know what this `RegisteredSub` is good for, i just replaced filter/watch with `EventSub`.

Reply:
OK. As noted above, the idea was that we can replace the complicated existing filter/watch logic with our new sub and not worry about merging past and future events. This requires understanding the context.